devito/dse/aliases.py

```diff
@@ -0,0 +1,249 @@
+            # - zero-offset if present, or
+            # - middle point rounded to the bottom
+            index = int((len(strides) - 1) / 2)
+            if (len(strides) - 1) % 2 == 0:
+                handle.append(strides[index])
+            elif 0 in strides:
+                handle.append(0)
+            else:
+                handle.append(strides[index])
+        COM.append(tuple(handle))
+
+    distances = []
+    for ofs in offsets:
+        handle = distance(COM, ofs)
+        assert len(handle) == 1
+        distances.append(handle.pop())
+
+    return COM, distances
+
+
+def calculate_offsets(indexeds):
+    """
+    Return a list of tuples, with one tuple for each indexed object appearing
+    in ``indexeds``. A tuple represents the offsets from the origin (0, 0, ..., 0)
+    along each dimension. All objects must use the same dimensions, in the same
+    order; otherwise, ``None`` is returned.
+
+    For example, given: ::
+
+        indexeds = [A[i,j,k], B[i,j+2,k+3]]
+
+    Return: ::
+
+        [(0, 0, 0), (0, 2, 3)]
+    """
+    processed = []
+    reference = indexeds[0].base.function.indices
+    for indexed in indexeds:
+        dimensions = indexed.base.function.indices
+        if dimensions != reference:
+            return None
+        handle = []
+        for d, i in zip(dimensions, indexed.indices):
+            offset = i - d
+            if offset.is_Number:
+                handle.append(offset)
+            else:
+                return None
+        processed.append(tuple(handle))
+    return tuple(reference), processed
+
+
+def distance(ofs1, ofs2):
+    """
+    Determine the distance of ``ofs2`` from ``ofs1``.
+    """
+    assert len(ofs1) == len(ofs2)
+    handle = set()
+    for o1, o2 in zip(ofs1, ofs2):
+        assert len(o1) == len(o2)
+        handle.add(tuple(i2 - i1 for i1, i2 in zip(o1, o2)))
+    return handle
+
+
+def is_translated(ofs1, ofs2):
+    """
+    Return True if ``ofs2`` is translated w.r.t. to ``ofs1``, False otherwise.
+
+    For example: ::
+
+        e1 = A[i,j] + A[i,j+1]
+        e2 = A[i+1,j] + A[i+1,j+1]
+
+    ``ofs1`` would be [(0, 0), (0, 1)], while ``ofs2`` would be [(1, 0), (1,1)], so
+    ``e2`` is translated w.r.t. ``e1`` by ``(1, 0)``, and True is returned.
+    """
+    return len(distance(ofs1, ofs2)) == 1
+
+
+def compare(e1, e2):
+    """
+    Return True if the two expressions e1 and e2 alias each other, False otherwise.
+    """
+    if type(e1) == type(e2) and len(e1.args) == len(e2.args):
+        if e1.is_Atom:
+            return True if e1 == e2 else False
+        elif isinstance(e1, Indexed) and isinstance(e2, Indexed):
+            return True if e1.base == e2.base else False
+        else:
+            for a1, a2 in zip(e1.args, e2.args):
+                if not compare(a1, a2):
+                    return False
+            return True
+    else:
+        return False
+
+
+class Alias(object):
+
```

Comment:
Docstring would be helpful, no? Actually that probably does extend to the routines below too.